File: ProblemSolve/1695.cpp
/*************************
 * Maximum Erasure Value *
 *************************/

class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int left = 0;
        int right = 0;
        int len = nums.size();
        int temp = 0;
        int ans = 0;
        
        //개수 세기, unique한 element인지 확인
        unordered_map<int,int> counts;
        
        while(right < len){
            counts[nums[right]]++;
            temp += nums[right];
            
            //포함 된 subarray가 unique하지 않은 경우
            //다시 unique하게 만들려고 window의 폭을 감소
            while(counts[nums[right]] > 1){
                counts[nums[left]]--;
                temp -= nums[left];
                left++;
            }
            
            //가장 큰 subarray의 sum 저장
            ans = max(ans, temp);
            right++;
        }
        
        return ans;
    }
};

File: ProblemSolve/1695.cs
/*************************
 * Maximum Erasure Value *
 *************************/

public class Solution {
    public int MaximumUniqueSubarray(int[] nums) {
        int[] hash = new int[10001];
        int ans = 0;
        int left = 0;
        int right = 0;
        int curSum = 0;
        
        while(right < nums.Length){
            hash[nums[right]]++;
            curSum += nums[right];
            
            while(hash[nums[right]] > 1){
                hash[nums[left]]--;
                curSum -= nums[left];
                left++;
            }
            
            ans = Math.Max(ans, curSum);
            right++;
        }
        
        return ans;
    }
}